ui/src/warp_runner/ui_adapter/mod.rs

```diff
@@ -0,0 +1,156 @@
+        warp_identity
+            .first()
+            .cloned()
+            .expect("could not get warp identity"),
+    )
+}
+
+pub async fn dids_to_identity(
+    dids: Vec<DID>,
+    account: &mut super::Account,
+) -> Vec<state::Identity> {
+    let mut ret = Vec::new();
+    for id in dids {
+        let ident = did_to_identity(id, account).await;
+        ret.push(ident);
+    }
+    ret
+}
+
+pub async fn conversation_to_chat(
+    conv: Conversation,
+    account: &mut super::Account,
+    messaging: &mut super::Messaging,
+) -> chats::Chat {
+    // todo: should Chat::participants include self?
+    let mut participants = Vec::new();
+    for id in conv.recipients() {
+        let ident = did_to_identity(id, account).await;
+        participants.push(ident);
+    }
+
+    let messages = match messaging
+        .get_messages(conv.id(), MessageOptions::default())
+        .await
+    {
+        Ok(m) => m,
+        Err(e) => match e {
+            warp::error::Error::EmptyMessage => vec![],
+            _ => {
+                println!("error: {}", e);
+                todo!("failed to get messages");
+            }
+        },
+    };
+
+    let unreads = messages.len() as u32;
+
+    chats::Chat {
+        id: conv.id(),
+        participants,
+        messages,
+        unreads,
+        replying_to: None,
+    }
+}
+
+// todo: put account and messaging in a module
+pub async fn convert_multipass_event(
+    event: warp::multipass::MultiPassEventKind,
+    account: &mut super::Account,
+    _messaging: &mut super::Messaging,
+) -> MultiPassEvent {
+    match event {
+        MultiPassEventKind::FriendRequestSent { to } => {
+            let identity = did_to_identity(to, account).await;
+            MultiPassEvent::FriendRequestSent(identity)
+        }
+        MultiPassEventKind::FriendRequestReceived { from } => {
+            let identity = did_to_identity(from, account).await;
+            //println!("friend request received: {:#?}", identity);
+            MultiPassEvent::FriendRequestReceived(identity)
+        }
+        MultiPassEventKind::IncomingFriendRequestClosed { .. }
+        | MultiPassEventKind::IncomingFriendRequestRejected { .. } => {
+            // can probably ignore this
+            MultiPassEvent::None
+        }
+        MultiPassEventKind::OutgoingFriendRequestClosed { did }
+        | MultiPassEventKind::OutgoingFriendRequestRejected { did } => {
+            let identity = did_to_identity(did, account).await;
+            MultiPassEvent::FriendRequestCancelled(identity)
+        }
+        MultiPassEventKind::FriendAdded { did } => {
+            let identity = did_to_identity(did, account).await;
+            MultiPassEvent::FriendAdded(identity)
+        }
+        MultiPassEventKind::FriendRemoved { did } => {
+            let identity = did_to_identity(did, account).await;
+            MultiPassEvent::FriendRemoved(identity)
+        }
+        MultiPassEventKind::IdentityOnline { did } => {
```

Comment:
Fine to use, but I wouldnt rely on this or `MultiPassEventKind::IdentityOffline` event all that much right now. It works, but sensitive (though I did a bit of refactoring that improves on it, but still subjected to things like timeouts, etc, which may trigger offline and online repeatedly - will improve on it though very soon).